### PR TITLE
CASMCMS-7677: Write new documentation for using BOS to apply rolling upgrades

### PR DESCRIPTION
**Pull request by rbak-hpe**
# Description

Adds new docs describing how BOS v2 can be used to perform rolling upgrades

# Checklist Before Merging

- [X] If I added any command snippets, the steps they belong to follow the prompt conventions (see [example][1]).
- [X] If I added a new directory, I also updated `.github/CODEOWNERS` with the corresponding team in [Cray-HPE][2].
- [X] My commits or Pull-Request Title contain my JIRA information, or I do not have a JIRA.

### CODE REVIEW

File: operations/boot_orchestration/Rolling_Upgrades.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
+# Rolling Upgrades using BOS
+
+> **`NOTE`** This section is for BOS V2 only.
+
+> **`NOTE`** This feature is the replacement for CRUS, which was deprecated in CSM 1.2.0.
+
+BOS V2 allows users to stage a boot artifacts, configuration and operation such as a reboot.
```
Comment (jsollom-hpe) [inline operations/boot_orchestration/Rolling_Upgrades.md:7]: Removed first 'a'.
Include Oxford comma.

File: operations/boot_orchestration/Rolling_Upgrades.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+The workload manager can later trigger the operation through BOS to apply that staged information, allowing rolling updates when nodes have no job running on them.
+
+## Workflow
+
+1) A sysadmin configures the workload manager to call the `applystaged` endpoint of BOS, along with sending a payload containing the `xname` of the components to be operated on.  For more information on the endpoint and payload see [Applying a Staged State](./Stage_Changes_with_BOS.md#applying-a-staged-state)
```
Comment (jsollom-hpe) [inline operations/boot_orchestration/Rolling_Upgrades.md:12]: `xname`s

File: operations/boot_orchestration/Rolling_Upgrades.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+    > **`NOTE`** The boot artifacts and configuration staged with BOS will not be applied if the node is rebooted outside BOS.
+    This is because BOS is caching the staged boot information and configuration internally, but not updating BSS and CFS until it immediately before to boots or reboots the nodes.
```
Comment (jsollom-hpe) [inline operations/boot_orchestration/Rolling_Upgrades.md:15]: Reword:
until immediately before it boots or reboots the nodes.

File: operations/boot_orchestration/Rolling_Upgrades.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+1) A sysadmin stages all of the boot information through BOS V2 by creating a session with the `staged` parameter.
+BOS will cache the boot artifacts and configuration and associate that information with the specified nodes. These nodes will not be booted or rebooted as a part of this staging.
+For more information on staging sessions, see [Creating a Staged Session](./Stage_Changes_with_BOS.md#creating-a-staged-session)
+
+1) The sysadmin indicates to the workload manager that a node reboot is needed.
+
+1) The workload manager triggers calls the `applystaged` endpoint for each node when they are ready.  BOS then copies the information staged for these components into their desired state, and BOS starts to operate on the nodes to bring them in line with their new desired state.
```
Comment (jsollom-hpe) [inline operations/boot_orchestration/Rolling_Upgrades.md:23]: Numerical disagreement:
for each node when it is ready 
OR 
for individual nodes when they are ready 

I am not sure if "to bring them in line with their new desired state." is too colloquial.
Potential substitution:
and attempts to make their actual states match their new desired states.

File: operations/boot_orchestration/Rolling_Upgrades.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+## Using staged sessions with Slurm
+
+Slurm can be configured to call a reboot script using the `RebootProgram` value in the `slurm.conf` file.
+Once this configuration is in place, and a staged session has been created, admins can issue a `scontrol reboot` command to Slurm.
```
Comment (jsollom-hpe) [inline operations/boot_orchestration/Rolling_Upgrades.md:28]: The first comma may be unnecessary, but you and I have had our share of comma contention, so take with a grain of salt.

File: operations/boot_orchestration/Rolling_Upgrades.md [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+BOS will cache the boot artifacts and configuration and associate that information with the specified nodes. These nodes will not be booted or rebooted as a part of this staging.
+For more information on staging sessions, see [Creating a Staged Session](./Stage_Changes_with_BOS.md#creating-a-staged-session)
+
+1) The sysadmin indicates to the workload manager that a node reboot is needed.
+
+1) The workload manager triggers calls the `applystaged` endpoint for each node when they are ready.  BOS then copies the information staged for these components into their desired state, and BOS starts to operate on the nodes to bring them in line with their new desired state.
+
+## Using staged sessions with Slurm
+
+Slurm can be configured to call a reboot script using the `RebootProgram` value in the `slurm.conf` file.
```
Comment (jsollom-hpe) [inline operations/boot_orchestration/Rolling_Upgrades.md:27]: I think we should a link to Slurm documentation for instructions on how to do this.

**Review by jsollom-hpe — CHANGES_REQUESTED** [review]

File: operations/boot_orchestration/Rolling_Upgrades.md (hunk 3 above)
Comment (rbak-hpe) [reply operations/boot_orchestration/Rolling_Upgrades.md:15]: wow, I'm not sure my previous sentence was even English.

File: operations/boot_orchestration/Rolling_Upgrades.md (hunk 6 above)
Comment (jsollom-hpe) [reply operations/boot_orchestration/Rolling_Upgrades.md:29]: Is that the only documentation that exists? Do our internal Slurm experts have anything? If not, then I think that is sufficient.

File: operations/boot_orchestration/Rolling_Upgrades.md [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    > **`NOTE`** The boot artifacts and configuration staged with BOS will not be applied if the node is rebooted outside BOS.
+    This is because BOS is caching the staged boot information and configuration internally, but not updating BSS and CFS until immediately before it boots or reboots the nodes.
+
+1) A sysadmin stages all of the boot information through BOS V2 by creating a session with the `staged` parameter.
+BOS will cache the boot artifacts and configuration and associate that information with the specified nodes. These nodes will not be booted or rebooted as a part of this staging.
+For more information on staging sessions, see [Creating a Staged Session](./Stage_Changes_with_BOS.md#creating-a-staged-session)
+
+1) The sysadmin indicates to the workload manager that a node reboot is needed.
+
+1) The workload manager triggers calls the `applystaged` endpoint for each node when it is ready.
```
Comment (jsollom-hpe) [inline operations/boot_orchestration/Rolling_Upgrades.md:24]: Re-word: triggers calls --> calls

**Review by jsollom-hpe — APPROVED** [review]

File: operations/boot_orchestration/Rolling_Upgrades.md (hunk 6 above)
Comment (rbak-hpe) [reply operations/boot_orchestration/Rolling_Upgrades.md:29]: I asked them to point me at documentation in the first place and I didn't get anything that wasn't in a private github repo, so I think this is the best we can do for now.

File: operations/boot_orchestration/Rolling_Upgrades.md [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+<!-- -->
+> **`NOTE`** This feature is the replacement for CRUS, which was deprecated in CSM 1.2.0.
+
+BOS V2 allows users to stage boot artifacts, configuration, and an operation such as a reboot.
+The workload manager can later trigger the operation through BOS to apply that staged information, allowing rolling updates when nodes have no job running on them.
+
+## Workflow
+
+1) A sysadmin configures the workload manager to call the `applystaged` endpoint of BOS, along with sending a payload containing the `xnames` of the components to be operated on.  For more information on the endpoint and payload see [Applying a Staged State](./Stage_Changes_with_BOS.md#applying-a-staged-state)
```
Comment (rustydb) [inline operations/boot_orchestration/Rolling_Upgrades.md:13]:
```suggestion
1. A sysadmin configures the workload manager to call the `applystaged` endpoint of BOS, along with sending a payload containing the `xnames` of the components to be operated on.  For more information on the endpoint and payload see [Applying a Staged State](./Stage_Changes_with_BOS.md#applying-a-staged-state)
```

File: operations/boot_orchestration/Rolling_Upgrades.md [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+    > **`NOTE`** The boot artifacts and configuration staged with BOS will not be applied if the node is rebooted outside BOS.
+    This is because BOS is caching the staged boot information and configuration internally, but not updating BSS and CFS until immediately before it boots or reboots the nodes.
+
+1) A sysadmin stages all of the boot information through BOS V2 by creating a session with the `staged` parameter.
```
Comment (rustydb) [inline operations/boot_orchestration/Rolling_Upgrades.md:18]:
```suggestion
1. A sysadmin stages all of the boot information through BOS V2 by creating a session with the `staged` parameter.
```

File: operations/boot_orchestration/Rolling_Upgrades.md [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+BOS will cache the boot artifacts and configuration and associate that information with the specified nodes. These nodes will not be booted or rebooted as a part of this staging.
+For more information on staging sessions, see [Creating a Staged Session](./Stage_Changes_with_BOS.md#creating-a-staged-session)
+
+1) The sysadmin indicates to the workload manager that a node reboot is needed.
```
Comment (rustydb) [inline operations/boot_orchestration/Rolling_Upgrades.md:22]:
```suggestion
1. The sysadmin indicates to the workload manager that a node reboot is needed.
```

File: operations/boot_orchestration/Rolling_Upgrades.md [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+
+1) The workload manager calls the `applystaged` endpoint for each node when it is ready.
```
Comment (rustydb) [inline operations/boot_orchestration/Rolling_Upgrades.md:24]:
```suggestion
1. The workload manager calls the `applystaged` endpoint for each node when it is ready.
```

**Review by rustydb — APPROVED** [review]

**Pull request merged**